### PR TITLE
[Feature] Show success message for late submissions and results

### PR DESCRIPTION
**Pull request by derLalla**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

If the exercise name has `BASE` or `SOLUTION` in its name, the student participation cannot be found during the result creation.
There is no indication for a student if he submits a programming exercise after the due date. So we want to have a success message for that.
Attempt to solve #2915 (?)

### Description

This fixes the bug with searching for a participation during the result creation. We now check if the build plan key ends with the string `-BASE` or `-SOLUTION` instead of just checking if the name exists within the string.

When a submission has no result and we are triggering a failed build within the Course Exercise Row or Course Exercise Detail, we are not using `lastGraded = true` anymore and are searching for any submission not only graded ones.

#### Test 1
1. Log in to Artemis
2. Participate in an exercise where the due date is over
3. Make changes and click on `Submit`
4. Directly afterwards change to the Course Exercise Detail and verify a green success notification which states that a late submission was received
5. Wait a few seconds and verify that another success notification shows that an `ungraded` result was received
6. At last the result history on the bottom of the page should have been updated with the latest result

#### Test 2
1. Log in to Artemis
2. Create a programming exercise where the name contains either BASE or SOLUTION
3. Participate as a student and verify that the build result is received as normal

### Test Coverage

course-exercise-details.component.ts: 80%

### CODE REVIEW

**Review by tiffca — DISMISSED** [review]
tested on ts1, the basic functionality worked hence approving, although ran into some strange behaviour. 

when initially navigating to the exercise after the due date I received this error "bad request" -> **not reproducible** 
GET https://artemistest.ase.in.tum.de/api/programming-exercise-participations/undefined/latest-pending-submission
image

on the course exercise details page it always displays "no graded result" even when there is a graded result 
image

image

I'm not sure if I received the second notification, although I do see the "not graded" button. Assuming that's what it's referring to. 
image


- tested in firefox and chrome

2nd test worked fine, created an exercise BASESOLUTION2925 that functioned as expected.

**Review by SimonEntholzer — DISMISSED** [review]
Testing worked fine locally, I got both messages:
image
unlike @tiffca I can see the graded result in the overview: 
image
But: I also get the same bad request in the same place.

Second test worked fine as expected, and code looks fine too :D

**Review by airelawaleria — APPROVED** [review]
all worked, but i've also encountered bad request issue as @tiffca

**Review by JohannesWeiss — DISMISSED** [review]
Tested on ts1 and encountered the same bad request as @tiffca. Aside from that, everything worked for me (both tests).

**Review by FrankeLukas — DISMISSED** [review]
Works, but something odd is going on. For some reason two submissions were generated instead of one:
grafik

Not sure if that is a bug in this PR though.

**Comment by derLalla** [comment]
> Works, but something odd is going on. For some reason two submissions were generated instead of one:
> grafik
> 
> Not sure if that is a bug in this PR though.

I think, if you would refresh the page, only one result would be shown. This might be a reference problem

**Comment by derLalla** [comment]
> Tested on ts1 and encountered the same bad request as @tiffca. Aside from that, everything worked for me (both tests).

I have fixed all of your problems regarding the `Bad Request` in the latest commit

**Review by tiffca — APPROVED** [review]
Core functionality works and the bad request no longer shows, hence approving. 

However I still see "No graded result" on the right in the exercise detail page. 
image

**Comment by derLalla** [comment]
> Core functionality works and the bad request no longer shows, hence approving.
> 
> However I still see "No graded result" on the right in the exercise detail page.
> image

This is intended, if you start an exercise after the due date. This view is only showing your graded result, also is the Course Exercise Row only showing your graded result. The online editor though, shows your ungraded result. So if you open it, you should see your ungraded result. The ungraded result can also be seen at the bottom of the Exercise Details.

File: src/test/javascript/spec/component/overview/exercise-details/course-exercise-details.component.spec.ts [hunk 1]
```diff
@@ -179,10 +189,18 @@ describe('CourseExerciseDetailsComponent', () => {
...
+
         const exerciseDetail = { ...exercise, studentParticipations: [studentParticipation] };
         const exerciseDetailReponse = of({ body: exerciseDetail });
 
+        // return initial participation for websocketService
+        stub(participationWebsocketService, 'getParticipationForExercise').returns(studentParticipation);
+
         mergeStudentParticipationStub.returns(studentParticipation);
+        const changedParticipation = cloneDeep(studentParticipation);
+        changedParticipation.results = [{ ...result, id: 2 }];
```
[the remark this answers is not in the document]
Comment (derLalla) [reply src/test/javascript/spec/component/overview/exercise-details/course-exercise-details.component.spec.ts]: fixed in f3c7d2f82103cdd32b5a6be16c9936d83738efd6

File: src/test/javascript/spec/component/overview/exercise-details/course-exercise-details.component.spec.ts [hunk 2]
```diff
@@ -156,6 +161,11 @@ describe('CourseExerciseDetailsComponent', () => {
                 // stub participationService, needed for team assignment
                 participationService = fixture.debugElement.injector.get(ParticipationService);
                 mergeStudentParticipationStub = stub(participationService, 'mergeStudentParticipations');
+
+                // stub participationService, needed for team assignment
+                participationWebsocketService = fixture.debugElement.injector.get(ParticipationWebsocketService);
+                subscribeForParticipationChangesStub = stub(participationWebsocketService, 'subscribeForParticipationChanges');
```
[the remark this answers is not in the document]
Comment (derLalla) [reply src/test/javascript/spec/component/overview/exercise-details/course-exercise-details.component.spec.ts:167]: fixed in f3c7d2f82103cdd32b5a6be16c9936d83738efd6

**Review by tiffca — APPROVED** [review]
Tested successfully

**Review by philgabr — APPROVED** [review]

**Review by airelawaleria — APPROVED** [review]
both test cases work

**Review by JohannesWeiss — APPROVED** [review]
Tested in testing session on ts1, re-approve.

**Review by LudwigWeide — APPROVED** [review]

**Review by simonzilker — APPROVED** [review]
Looks good

**Review by EneaGore — APPROVED** [review]

**Review by marvinchimeka — APPROVED** [review]
Works as expected👍

**Review by JannisBittlmayer — APPROVED** [review]
Tested with @FlorianStang, works as intended.

**Review by FlorianStang — APPROVED** [review]
Tested on TS1 with @JannisBittlmayer, everything worked as intended.

**Review by FrankeLukas — APPROVED** [review]
Re-approve.

**Review by SimonEntholzer — APPROVED** [review]
Re-approve

**Review by FuchsDominik — APPROVED** [review]
Code looks also good :)

**Review by ivanchimeno — APPROVED** [review]
Tested and it works as intended 👍

**Review by iozbeyli — APPROVED** [review]
Code looks good

**Pull request merged**